indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java

```diff
@@ -619,40 +631,54 @@ private boolean tryAssignTask(final Task task, final RemoteTaskRunnerWorkItem ta
       } else {
         strategy = workerConfig.getSelectStrategy();
       }
-      final Optional<ImmutableZkWorker> immutableZkWorker = strategy.findWorkerForTask(
-          config,
-          ImmutableMap.copyOf(
-              Maps.transformEntries(
-                  Maps.filterEntries(
-                      zkWorkers, new Predicate<Map.Entry<String, ZkWorker>>()
-                      {
-                        @Override
-                        public boolean apply(Map.Entry<String, ZkWorker> input)
+
+      ZkWorker assignedWorker = null;
+      try {
+        final Optional<ImmutableZkWorker> immutableZkWorker = strategy.findWorkerForTask(
+            config,
+            ImmutableMap.copyOf(
+                Maps.transformEntries(
+                    Maps.filterEntries(
+                        zkWorkers, new Predicate<Map.Entry<String, ZkWorker>>()
                         {
-                          return !lazyWorkers.containsKey(input.getKey());
+                          @Override
+                          public boolean apply(Map.Entry<String, ZkWorker> input)
+                          {
+                            return !lazyWorkers.containsKey(input.getKey()) &&
+                                   !workersWithUnacknowledgedTask.containsKey(input.getKey());
+                          }
                         }
-                      }
-                  ),
-                  new Maps.EntryTransformer<String, ZkWorker, ImmutableZkWorker>()
-                  {
-                    @Override
-                    public ImmutableZkWorker transformEntry(
-                        String key, ZkWorker value
-                    )
+                    ),
+                    new Maps.EntryTransformer<String, ZkWorker, ImmutableZkWorker>()
                     {
-                      return value.toImmutable();
+                      @Override
+                      public ImmutableZkWorker transformEntry(
+                          String key, ZkWorker value
+                      )
+                      {
+                        return value.toImmutable();
+                      }
                     }
-                  }
-              )
-          ),
-          task
-      );
-      if (immutableZkWorker.isPresent()) {
-        final ZkWorker zkWorker = zkWorkers.get(immutableZkWorker.get().getWorker().getHost());
-        return announceTask(task, zkWorker, taskRunnerWorkItem);
-      } else {
+                )
+            ),
+            task
+        );
+
+        if (immutableZkWorker.isPresent()
+            &&
+            workersWithUnacknowledgedTask.putIfAbsent(immutableZkWorker.get().getWorker().getHost(), task.getId())
+            == null) {
```

Comment:
I'm thinking of the case where we have 1 worker and 2 tasks, and runPendingTasks was triggered twice, then I believe this could happen:

- thread 1 assigns task 1 to worker 1
- thread 2 tries to assign task 2 to worker 1, but cannot because it has not yet acknowledged task 1
- now task 2 will not be assigned by either thread, and will stay pending even though there is some capacity